torch/_subclasses/fake_tensor.py

```diff
@@ -1414,6 +1524,215 @@ def __exit__(self, a, b, c):
             if maybe_prev_fake_mode is not None:
                 torch._C._set_dispatch_mode(maybe_prev_fake_mode)
 
+    @classmethod
+    def cache_info(cls) -> DispatchCacheInfo:
+        """
+        Query the state of the dispatch cache.
+        """
+        return DispatchCacheInfo(
+            FakeTensorMode.cache_hits,
+            FakeTensorMode.cache_misses,
+            dict(FakeTensorMode.cache_bypasses),
+            len(FakeTensorMode.cache),
+        )
+
+    @classmethod
+    def cache_clear(cls):
+        """
+        Clear the dispatch cache.
+        """
+        cls.hits = 0
+        cls.misses = 0
+        cls.cache_bypasses.clear()
+        cls.cache.clear()
+
+    def _cached_dispatch_impl(
+        self,
+        func: OpOverload,
+        types: Tuple[Any],
+        args: Tuple[Any],
+        kwargs: Dict[str, Any],
+    ):
+        """
+        Lookup a cache entry for the given arguments. If none exists, dispatch
+        and cache the result (if the result is cache-able).
+        """
+        try:
+            key = self._cache_key(func, args, kwargs)
+        except _BypassDispatchCache as e:
+            FakeTensorMode.cache_bypasses[e.reason] += 1
+            return self._dispatch_impl(func, types, args, kwargs)
+
+        entry = FakeTensorMode.cache.get(key, None)
+        if entry is not None:
+            FakeTensorMode.cache_hits += 1
+            output = self._output_from_cache_entry(entry, func, args)
+        else:
+            FakeTensorMode.cache_misses += 1
+            output = self._dispatch_impl(func, types, args, kwargs)
+            try:
+                self._create_cache_entry(key, args, output)
+            except _BypassDispatchCache as e:
+                FakeTensorMode.cache_bypasses[e.reason] += 1
+
+        return output
+
+    def _cache_key(
+        self, func: OpOverload, args: Tuple[Any], kwargs: Dict[str, Any]
+    ) -> _DispatchCacheKey:
+        """
+        Create a cache key given the dispatch args. Raise _BypassDispatchCache
+        for any situation that precludes caching.
+        """
+        # Avoid caching for any ops that would require a more sophisticated
+        # caching implementation, e.g., data dependent ops or ops that modify
+        # the inputs.
+        if torch.Tag.data_dependent_output in func.tags:
```

Comment:
This is the random list of "if special-case-x then don't cache" cases I found. It makes me worried that I won't have found every possibility. It's going to be quite confusing for users if we ever serve a cache hit that would actually produce a different result if actually evaluated.

Reply:
Yea, that is concerning. We should have a mode of running comparing cached results with a re-run, and see if the metadata differs. 

cc @bdhirsh @ezyang @voznesenskym  for thoughts